README.md
# TYPO3-Direct-Request
This is a TYPO3-Extension to fake an HTTP request and perform a direct request to the frontend triggered by the backend.

## Download / Installation

You can download and install this extension or use composer.

## Copyright / License

Copyright: (c) 2010 - 2019, AOE GmbH
License: GPLv3, <http://www.gnu.org/licenses/gpl-3.0.en.html>

## Contributing

	1. Fork the repository on Github
	2. Create a named feature / bugfix branch (like `feature_add_something_new` or `bugfix\thing_which_does_not_work`)
	3. Write your change
	4. Write tests for your change (if applicable)
	5. Run the tests, ensuring they all pass
	6. Submit a Pull Request using Github


ChangeLog
2010-07-14  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Headers are not used on executing a direct request
	* Fixed bug: PHP notice is shown if no HTTP headers are given
	* Raised version number to 0.0.3

2010-05-28  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Undefined index if URL does not have a query part
	* Raised version number to 0.0.2

2010-05-12  Oliver Hader  <oliver@typo3.org>

	* Fixed bug: Wrong definitions in ext_localconf.php file and ext_tables.php is superfluous

2010-05-05  Oliver Hader  <oliver@typo3.org>

	* Raised version number to 0.0.1

2010-04-30  Oliver Hader  <oliver@typo3.org>

	* Kickstarted initial version (thanks to Fabrizio Branca for providing most of the features in the crawler extension)

cli/request.php
<?php


/**
 * Retrieve path (taken from cli_dispatch.phpsh)
 */
// echo realpath(dirname(__FILE__).'/../../../..'), "\n";


	// Get path to this script
$temp_PATH_thisScript = isset($_SERVER['argv'][0]) ? $_SERVER['argv'][0] : (isset($_ENV['_']) ? $_ENV['_'] : $_SERVER['_']);

	// Figure out if the path is relative
$relativePath = FALSE;
if (stristr(PHP_OS,'win') && !stristr(PHP_OS,'darwin')) {
		// Windows
	if (!preg_match('/^([A-Z]:)?\\\/', $temp_PATH_thisScript)) {
		$relativePath = TRUE;
	}
} else {
		// *nix, et al
	if ($temp_PATH_thisScript{0} != '/') {
		$relativePath = TRUE;
	}
}

	// Resolve path
if ($relativePath) {
	$workingDirectory = $_SERVER['PWD'] ? $_SERVER['PWD'] : getcwd();
	if ($workingDirectory) {
		$temp_PATH_thisScript =
			$workingDirectory.'/'.preg_replace('/\.\//','',$temp_PATH_thisScript);
		if (!@is_file($temp_PATH_thisScript)) {
			die ('Relative path found, but an error occured during resolving of the absolute path: '.$temp_PATH_thisScript.chr(10));
		}
	} else {
		die ('Relative path found, but resolving absolute path is not supported on this platform.'.chr(10));
	}
}

$typo3Root = preg_replace('#typo3conf/ext/directrequest/cli/request.php$#', '', $temp_PATH_thisScript);



/**
 * Second paramater is a base64 encoded serialzed array of header data
 */
if (isset($_SERVER['argv'][3])) {
	$additionalHeaders = unserialize(base64_decode($_SERVER['argv'][3]));
	if (is_array($additionalHeaders)) {
		foreach ($additionalHeaders as $additionalHeader) {
			if (strpos($additionalHeader, ':') !== false) {
				list($key, $value) = explode(':', $additionalHeader, 2);
				$key = str_replace('-', '_', strtoupper(trim($key)));
				if ($key != 'HOST') {
					$_SERVER['HTTP_' . $key] = trim($value);
				}
			}
		}
	}
}

// put parsed query parts into $_GET array
$urlParts = parse_url($_SERVER['argv'][2]);

// Populating $_GET and $_REQUEST is query part is set:
if (isset($urlParts['query'])) {
	parse_str($urlParts['query'], $_GET);
	parse_str($urlParts['query'], $_REQUEST);
}

// Populating $_POST
$_POST = array();
// Populating $_COOKIE
$_COOKIE = array();

// Get the TYPO3_SITE_PATH of the website frontend:
$typo3SitePath = $_SERVER['argv'][1];

// faking the environment
$_SERVER['ORIG_SCRIPT_FILENAME'] = '';
$_SERVER['DOCUMENT_ROOT'] = preg_replace('#' . preg_quote($typo3SitePath, '#') . '$#', '', $typo3Root);
$_SERVER['HTTP_USER_AGENT'] = 'CLI Mode';
$_SERVER['HTTP_HOST'] = $_SERVER['SERVER_NAME'] = $urlParts['host'];
$_SERVER['SCRIPT_NAME'] = $_SERVER['PHP_SELF'] = $typo3SitePath . 'index.php';
$_SERVER['SCRIPT_FILENAME'] = $_SERVER['PATH_TRANSLATED'] = $typo3Root . 'index.php';
$_SERVER['QUERY_STRING'] = (isset($urlParts['query']) ? $urlParts['query'] : '');
$_SERVER['REQUEST_URI'] = $urlParts['path'] . (isset($urlParts['query']) ? '?' . $urlParts['query'] : '');

// Define a port if used in the URL: 
if (isset($urlParts['port'])) {
	$_SERVER['HTTP_HOST'] .= ':' . $urlParts['port'];
	$_SERVER['SERVER_PORT'] = $urlParts['port'];
}
// Define HTTPS disposal:
if ($urlParts['scheme'] === 'https') {
	$_SERVER['HTTPS'] = 'on';
}

chdir($typo3Root);
include($typo3Root . '/index.php');

?>

classes/class.tx_directrequest_manager.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2010 AOE media (dev@aoemedia.de)
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

/**
 * Class that manages direct requests.
 */
class tx_directrequest_manager implements t3lib_Singleton {
	/**
	 * @var array
	 */
	protected $extensionSettings;

	/**
	 * Constructs this object.
	 */
	public function __construct() {
		$this->extensionSettings = (array) unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['directrequest']);
	}

	/**
	 * Executes a direct request.
	 *
	 * @param string $url The full qualified URL to be requested
	 * @param array $headers Additions HTTP headers to transferred along
	 * @return array The result with contents, response and request headers
	 */
	public function execute($url, array $headers = NULL) {
		$command = $this->getShellCommand($url, $headers);
		$content = $this->executeShellCommand($command);

		$result = array(
			'content' => $content,
			'request' => '',
			'response' => '',
		);

		return $result;
	}

	/**
	 * Generates the shell command that is executed to perform the request.
	 *
	 * @param string $url The full qualified URL to be requested
	 * @param array $headers Additions HTTP headers to transferred along
	 * @return string The command string
	 */
	protected function getShellCommand($url, array $headers = NULL) {
		$commandParts = array(
			escapeshellcmd($this->getPhpPath()),
			escapeshellarg(t3lib_extMgm::extPath('directrequest') . 'cli/request.php'),
			escapeshellarg($this->getFrontendBasePath()),
			escapeshellarg($url),
			escapeshellarg(base64_encode(serialize($headers))),
		);

		$command = trim(implode(' ', $commandParts));

		return $command;
	}

	/**
	 * Executes a shell command and returns the outputted result.
	 *
	 * @param string $command Shell command to be executed
	 * @return string Outputted result of the command execution
	 */
	protected function executeShellCommand($command) {
		$result = shell_exec($command);
		return $result;
	}

	/**
	 * Gets the path to the PHP executable.
	 *
	 * @return string Path to the PHP executable
	 */
	protected function getPhpPath() {
		$phpPath = 'php';

		if (isset($this->extensionSettings['phpPath']) && $this->extensionSettings['phpPath']) {
			$phpPath = $this->extensionSettings['phpPath'];
		}

		return $phpPath;
	}

	/**
	 * Gets the base path of the website frontend.
	 * (e.g. if you call http://mydomain.com/cms/index.php in
	 * the browser the base path is "/cms/")
	 *
	 * @return string Base path of the website frontend
	 */
	protected function getFrontendBasePath() {
		$frontendBasePath = '/';

		// Get the path from the extension settings:
		if (isset($this->extensionSettings['frontendBasePath']) && $this->extensionSettings['frontendBasePath']) {
			$frontendBasePath = $this->extensionSettings['frontendBasePath'];
		// If not in CLI mode the base path can be determined from $_SERVER environment:
		} elseif (!defined('TYPO3_cliMode') || !TYPO3_cliMode) {
			t3lib_div::getIndpEnv('TYPO3_SITE_PATH');
		}

		// Base path must be '/<pathSegements>/':
		if ($frontendBasePath != '/') {
			$frontendBasePath = '/' . ltrim($frontendBasePath, '/');
			$frontendBasePath = rtrim($frontendBasePath, '/') . '/';
		}

		return $frontendBasePath;
	}
}

ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "directrequest".
#
# Auto generated 05-05-2010 13:22
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Direct Request',
	'description' => 'Extension the fake an HTTP request and perform a direct request to the frontend triggered by the backend',
	'category' => 'misc',
	'author' => 'AOE media',
	'author_email' => 'dev@aoemedia.de',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'beta',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => 'AOE media GmbH',
	'version' => '0.0.3',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.3.0-0.0.0',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:7:{s:9:"ChangeLog";s:4:"bd3d";s:16:"ext_autoload.php";s:4:"e869";s:21:"ext_conf_template.txt";s:4:"d67d";s:17:"ext_localconf.php";s:4:"ae1f";s:14:"ext_tables.php";s:4:"9c67";s:42:"classes/class.tx_directrequest_manager.php";s:4:"f92b";s:15:"cli/request.php";s:4:"0bdb";}',
	'suggests' => array(
	),
);

?>